Documentation/boot-flow.md

```diff
@@ -0,0 +1,56 @@
+# Boot Flow
+
+## Storage
+
+Firmware resides in a storage part. One common storage to use is SPI flash,
+commonly NOR, but other parts are possible as well, depending on the platform.
+Those may be SPI NAND, which can hold larger firmware images, eMMC/MMC, SD card,
+USB drives, and in split scenarios, a fraction may even reside on the network.
+
+For simplicity, we describe a typical NOR flash for the example.
+
+## Stages
+
+Due to the nature of platforms, firmware needs to start small. From stage to
+stage, we set up an environment such that an operating system may be booted or
+any other payload executed.
+With oreboot, we focus on embedding [LinuxBoot](https://linuxboot.org).
+
+![oreboot flash and flow](img/firmware-matryoshka.png)
+
+### Stage 1: SRAM/XIP
+
+In the first stage, we initialize the main clocks, some GPIOs, a serial port,
+and potentially light up LEDs to get early feedback. Then we initialize the DRAM
+controller and make the large DRAM available. The only memory we have up to that
+point is SRAM plus registers. Depending on the platform, [XIP (eXecute in-place)
+](https://en.wikipedia.org/wiki/Execute_in_place) may be possible, e.g., when
+the boot storage (NOR flash) is mapped to memory.
+
+The final step in this early stage is to copy the next stage to DRAM and run it.
+
+### Stage 2: DRAM
+
+In the second stage, we are all set, and depending on the platform, we could
+already run the final OS, a hypervisor, or any other payload. However, some
+platforms require extra steps. For this example, we will look at RISC-V.
+
+RISC-V describes [SBI](https://github.com/riscv-non-isa/riscv-sbi-doc), the
+Supervisor Binary Interface, in the [privileged spec](
+it://riscv.org/technical/specifications/). The role of SBI is to set up and/or
+delegate exception and interrupt handling, register functions for the OS to call
+into, and drop into a lower privileged mode named S-mode which makes an [MMU](
+https://en.wikipedia.org/wiki/Memory_management_unit) available, and finally
+execute the operating system. This is where we hand over to LinuxBoot, our
+boot loader environment. In oreboot, we implement SBI via [RustSBI](
+https://github.com/rustsbi/rustsbi), a Rust crate offering all the necessary
+functions we need as a library.
+
+### Stage 3: Payload (LinuxBoot)
+
+In this final stage, we are presented with a Linux environment. To have all the
+familiar commands, such as `ls`, `cat`, etc, as well as boot loaders, we choose
+[u-root](https://u-root.org) to build our initramfs. However, you may as well
+embed [cpud](https://github.com/u-root/cpu) directly, or your own custom app.
```

Comment:
do we have a network setup yet or can we say something about network? These boards are pretty useless without it.

Reply:
Well, it "depends on the board". Some have Wi-Fi, some have ethernet, some may be extended, e.g., with a SPI ethernet or SDIO Wi-Fi module.

Could you maybe elaborate on how you picture to "have a network setup"? I have many different network setups in the above sense of "it depends".